clients/drcachesim/tools/filter/record_filter.cpp

```diff
@@ -345,13 +400,61 @@ record_filter_t::parallel_shard_error(void *shard_data)
 bool
 record_filter_t::write_trace_entry(per_shard_t *shard, const trace_entry_t &entry)
 {
+    if (shard->output_entry_count == 0 && entry.type != TRACE_TYPE_HEADER) {
+        // When core-sharded with initially-idle cores we can start without a header.
+        // XXX i#6703: The scheduler should insert these headers for us.
+        std::vector<trace_entry_t> header;
+        trace_entry_t record = {};
+        record.type = TRACE_TYPE_HEADER;
+        // Our own stream's version + filetype are 0 so we use another shard's.
+        std::lock_guard<std::mutex> guard(input_info_mutex_);
+        record.addr = static_cast<addr_t>(version_);
+        header.push_back(record);
+        record.type = TRACE_TYPE_MARKER;
+        record.size = TRACE_MARKER_TYPE_VERSION;
+        record.addr = static_cast<addr_t>(version_);
+        header.push_back(record);
+        record.type = TRACE_TYPE_MARKER;
+        record.size = TRACE_MARKER_TYPE_FILETYPE;
+        record.addr = static_cast<addr_t>(filetype_);
+        header.push_back(record);
+        // file_reader_t::open_input_file demands tid+pid so we insert sentinel values.
+        // We can't use INVALID_THREAD_ID as scheduler_t::open_reader() loops until
+        // record_type_has_tid() which requires record.marker.tid != INVALID_THREAD_ID.
+        record.type = TRACE_TYPE_THREAD;
+        record.size = sizeof(thread_id_t);
+        record.addr = static_cast<addr_t>(IDLE_THREAD_ID);
```

Comment:
IDLE_THREAD_ID should document that it is part of the trace format. In fact, to make it easier to track compatibility, we should have all stored trace format related constants at one place.

Reply:
There is a #6703 comment asking to move this to scheduler.h at line 63 in utils.h.